EditStateUseCase with Complete Test Coverage

Hasa9nAli opened the pull request:
## Description
This PR implements the `EditStateUseCase` class with full test coverage (100% method and class coverage) as verified by the coverage screenshots.

## Changes
- Implemented `EditStateUseCase` with all required validation logic
- Added comprehensive unit tests covering all scenarios
- Achieved 100% test coverage for both methods and class

## Validation Logic
The use case validates:
1. State title cannot be blank
2. State must belong to an existing project

## Test Coverage
All methods are thoroughly tested including:
- Main `invoke` method scenarios
- `isValidState` validation
- `validateTitle` with blank and non-blank cases
- `validateProject` with existing and non-existing projects
- Default `isProjectExist` behavior

## Screenshots
<img width="791" alt="Screenshot 2025-04-29 at 2 06 16 PM" src="https://github.com/user-attachments/assets/d0da585d-89a6-4e95-87db-8ca848f82d44" />


## Testing
 tests:
- `invoke should return same state when validation passes`
- `invoke should throw when title is blank`
- `invoke should throw when project doesn't exist`
- `isValidState should pass with valid state`
- `validateTitle should throw for blank title`
- `validateTitle should pass for non-blank title`
- `validateProject should throw for non-existent project`
- `validateProject should pass for existing project`
- `isProjectExist should return false by default`

Hasa9nAli committed:
init EditStateUseCase without implementation

Hasa9nAli committed:
add EditStateUseCaseTest with coverage 100%

Hasa9nAli committed:
add DeleteStateUseCase without implementation of method for testing purpose

Hasa9nAli committed:
add test cases in DeleteStateUseCaseTest with archive 100% coverage

ZeyadAbdullah679 commented on src/main/kotlin/logic/usecases/EditStateUseCase.kt:
```diff
@@ -0,0 +1,32 @@
+
+class EditStateUseCase {
+
+    operator fun invoke(state: State): State{
```
You can encapsulate your state as changing the id would result in unexpected results, so you can use string parameter indicates the name of the state

ZeyadAbdullah679 commented on src/main/kotlin/logic/usecases/EditStateUseCase.kt:
```diff
@@ -0,0 +1,32 @@
+        validateProject(state.projectId)
+    }
+
+    private fun validateTitle(title: String) {
```
This is the only function you will need while checking the title

ZeyadAbdullah679 commented on src/test/kotlin/EditStateUseCaseTest.kt:
```diff
@@ -0,0 +1,152 @@
+
+    @BeforeEach
+    fun setup() {
+        editStateUseCase = spyk(EditStateUseCase())
```
you should use mockk instead of spyk as it is not a case you have to use a spy here mockk is enough for that.

ZeyadAbdullah679 commented on src/test/kotlin/EditStateUseCaseTest.kt:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    @Test
+    fun `invoke should return same state when validation passes`() {
```
keep it without invoke at first

AimanYosofiUlt replied:
yes to keep with should_when convention, i think it's good practice 

ZeyadAbdullah679 commented on src/test/kotlin/EditStateUseCaseTest.kt:
```diff
@@ -0,0 +1,152 @@
+            title = "Valid Title",
+            projectId = UUID.randomUUID()
+        )
+        mockPrivateMethod("isProjectExist", validState.projectId, true)
```
Same logic as the previous branch

AimanYosofiUlt commented on src/test/kotlin/EditStateUseCaseTest.kt:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    @Test
+    fun `invoke should throw when title is blank`() {
```
Can you be more specific about what the exception is, and I see you use IllegalArgumentException I suggest use custom exception

AimanYosofiUlt reviewed (changes requested):
Just take a moment to check my reviews—together, we can keep improving our work

Hasa9nAli committed:
add updateStateTitleForSpecificProjectById to ProjectsRepository

Hasa9nAli committed:
add IllegalTitle to handel add invalid title

Hasa9nAli committed:
add projectRepository and remove unwanted validate method

Hasa9nAli committed:
remove unwanted test cases, refactoring name of function, and use mockk repo and testing

Hasa9nAli committed:
add StateNotExistException

Hasa9nAli committed:
add in ProjectsRepository
1- deleteState
2- getStateById
3- getStatesById

Hasa9nAli committed:
add project repo to constructor of DeleteStateUseCase and add projectId to invoke method

Hasa9nAli committed:
re-test of DeleteStateUseCaseTest

Moha-ALjabouri commented on src/main/kotlin/logic/usecases/EditStateUseCase.kt:
```diff
@@ -0,0 +1,20 @@
+        return state.copy(title = title) // fake for now
+    }
+
+    private fun isValidTitle(title: String) {
```
I think you need call isValidTitle function inside invoke function

Hasa9nAli replied:
yes, that help to make 100 coverage in the method, but i will call it when add correct implementation to archive 100 coverage of methods ❤️

Moha-ALjabouri commented on src/test/kotlin/EditStateUseCaseTest.kt:
```diff
@@ -0,0 +1,76 @@
+        val exception = assertThrows<IllegalTitle> {
+            editStateUseCase(invalidState, invalidState.title)
+        }
+        assertEquals("State title cannot be blank", exception.message)
```
I suggest use  Truth.assertThat insteat of assertEqual

Hasa9nAli replied:
i will do that ❤️

AimanYosofiUlt reviewed (dismissed):
Nice work

ZeyadAbdullah679 reviewed (changes requested):
Kindly, update you code according to the comments and resolve the conflicts by merging development branch here

ZeyadAbdullah679 commented on src/main/kotlin/logic/usecases/EditStateUseCase.kt:
```diff
@@ -0,0 +1,20 @@
+    private val projectsRepository: ProjectsRepository
+){
+
+    operator fun invoke(state: State, title: String): State {
```
I think you don't have to return the state a boolean value indicating if the change done or not would be enough

Ahmedsayed0895 replied:
@ZeyadAbdullah679 i agree with you but from another perspective maybe it be better for UI to have the model after editing and take any wanted attribute? 
what do you think?

ZeyadAbdullah679 replied:
@Ahmedsayed0895 I don't think so. If we want to check whether the operation was successful, we need to rely on the returned state title or something similar. Returning a Boolean can help determine what to show in the UI—for example, if the state was edited successfully, show one message; otherwise, show a different one or retry the operation.

Ahmedsayed0895 replied:
@ZeyadAbdullah679 Ok i got you, it's clear for me now .. thanks

Hasa9nAli replied:
@ZeyadAbdullah679 in interview Bareq say when you edit the something you should return the updated value , and from another perspective when i work with api when i use edit end point the back end developer always return the updated item

ZeyadAbdullah679 commented on src/main/kotlin/data/repositories/ProjectsRepositoryImpl.kt:
```diff
@@ -1,5 +1,10 @@
-class ProjectsRepositoryImpl: ProjectsRepository
\ No newline at end of file
+class ProjectsRepositoryImpl: ProjectsRepository {
+    override fun updateStateTitleForSpecificProjectById(projectId: UUID, stateId: UUID) {
```
Overly verbose function name.

Hasa9nAli replied:
depend on meaning full concept on clean code there is not problem if the name be long

ZeyadAbdullah679 commented on src/main/kotlin/data/repositories/ProjectsRepositoryImpl.kt:
```diff
@@ -1,5 +1,10 @@
 
-class ProjectsRepositoryImpl: ProjectsRepository
\ No newline at end of file
+class ProjectsRepositoryImpl: ProjectsRepository {
```
Be careful that `ProjectRepository` is responsible only for project related operations, so updating the state here's violating the SRP principle. You can do this in new `StatesRepository`.

Hasa9nAli replied:
Done

ZeyadAbdullah679 commented on src/main/kotlin/utilities/Exception.kt:
```diff
@@ -0,0 +1,3 @@
+package utilities
+
+class IllegalTitle(message: String): Exception(message)
```
Consider changing the name to something meaningful like IllegalStateTitle

Hasa9nAli replied:
i think as you said, but i use that to be global on all title validation like (project, task, and state)

Hasa9nAli replied:
Done

ZeyadAbdullah679 commented on src/test/kotlin/EditStateUseCaseTest.kt:
```diff
@@ -0,0 +1,76 @@
+
+class EditStateUseCaseTest {
+    private lateinit var editStateUseCase: EditStateUseCase
+    private lateinit var projectsRepository: ProjectsRepository
```
Consider doing the logic related to the states in different repository like `StatesRepository`

Hasa9nAli replied:
i will do that ❤️

Hasa9nAli replied:
Done

ZeyadAbdullah679 commented on src/test/kotlin/EditStateUseCaseTest.kt:
```diff
@@ -0,0 +1,76 @@
+        )
+        val newTitle = "Updated State"
+
+        every { projectsRepository.updateStateTitleForSpecificProjectById(any(), any()) }
```
every should have a return value.

Ahmedsayed0895 commented on src/test/kotlin/EditStateUseCaseTest.kt:
```diff
@@ -0,0 +1,76 @@
+        )
+
+        // When & Then
+        val exception = assertThrows<IllegalTitle> {
```
Great work, Hasan! Just a little note here .. the test case name contains the exception name, "IllegalStateTitle," while the assertThrows called as"IllegalTitle."

Hasa9nAli replied:
ohhh thank you for your note

Ahmedsayed0895 reviewed (commented):
Great work and great effort just consider my little note

Hasa9nAli committed:
add StateRepository that handle
1- getStateById
2- getStates
3- getStatesByProject
4- updateStateTitle
5- addState
6- deleteState

Hasa9nAli committed:
rename exception from IllegalTitle to IllegalStateTitle

Hasa9nAli committed:
inject stateRepository to editStateUseCase

Hasa9nAli committed:
replace ProjectsRepository to stateRepository

Hasa9nAli committed:
remove body of ProjectsRepository

Hasa9nAli committed:
Merge branch 'development' into test/edit-state-test

Hasa9nAli committed:
resolve conflicts

Hasa9nAli committed:
Merge branch 'development' into test/edit-state-test

Hasa9nAli committed:
resolve conflicts

AimanYosofiUlt reviewed (dismissed):
Thanks for the changes

Hasa9nAli committed:
resolve conflicts

Hasa9nAli committed:
resolve conflicts

Hasa9nAli committed:
update assert and make return valid

Hasa9nAli committed:
resolve conflicts

AimanYosofiUlt reviewed (dismissed):
Nice work

ZeyadAbdullah679 commented on src/main/kotlin/data/repositories/ProjectsRepositoryImpl.kt:
```diff
@@ -1,7 +1,11 @@
+import java.util.*
 
 class ProjectsRepositoryImpl: ProjectsRepository{
+    override fun isExist(projectId: UUID): Boolean {
```
Responsibility of another teammates

ZeyadAbdullah679 commented on src/main/kotlin/logic/repositoies/ProjectsRepository.kt:
```diff
@@ -1,5 +1,8 @@
+import java.util.UUID
+
 interface ProjectsRepository {
+    fun isExist(projectId: UUID): Boolean
```
Responsibility of another teammates

ZeyadAbdullah679 commented on src/main/kotlin/data/repositories/StateRepositoryImp.kt:
```diff
@@ -0,0 +1,32 @@
+import logic.repositoies.StateRepository
+import java.util.*
+
+class StateRepositoryImp: StateRepository {
```
StateRepositoryImpl not StateRepositoryImp

ZeyadAbdullah679 commented on src/main/kotlin/logic/usecases/AddStateUseCase.kt:
```diff
@@ -0,0 +1,18 @@
+import logic.repositoies.ProjectsRepository
+
+class AddStateUseCase(
+    private val projectRepository: ProjectsRepository,
```
Inject only statesRepository and use add states from there

ZeyadAbdullah679 replied:
if you need get project by id inject projects repository in the states repository implementation

ZeyadAbdullah679 commented on src/main/kotlin/data/repositories/StateRepositoryImp.kt:
```diff
@@ -0,0 +1,32 @@
+        return emptyList() // fake to TDD
+    }
+
+    override fun getStatesByProject(projectId: UUID): List<TaskState> {
```
Named it like the one in the task management in our plan.
getByProjectId

Hasa9nAli committed:
resolve conflicts

Hasa9nAli committed:
add deleteState and getStateById to StatesRepository and call it in DeleteStateUseCase

Hasa9nAli committed:
resolve conflict

Hasa9nAli committed:
replace ProjectRepository to StateRepository

Hasa9nAli committed:
add StateRepository to handel state management

Hasa9nAli committed:
add StateNotExistException

Hasa9nAli committed:
refactor EditStateUseCase

Hasa9nAli committed:
refactor DeleteStateUseCase

Hasa9nAli commented:
that merged with state management pull request
